eppo_core/src/eval/eval_assignment.rs

```diff
@@ -91,6 +91,56 @@ pub fn get_assignment_details(
     (result_with_details, event)
 }
 
+pub fn get_subject_assignments(
+    configuration: Option<&Configuration>,
+    subject_key: &Str,
+    subject_attributes: &Arc<Attributes>,
+    now: DateTime<Utc>,
+) -> HashMap<String, Result<Option<Assignment>, EvaluationError>> {
+    get_assignments_with_visitor(
+        configuration,
+        &mut NoopEvalVisitor,
+        subject_key,
+        subject_attributes,
+        now,
+    )
+}
+
+/**
+ * Evaluate all flags for the given subject and return a map of flag keys to assignments.
+ *
+ * If an error occurs while evaluating a flag,
+ * it is returned for the resulting flag key.
+ */
+pub(super) fn get_assignments_with_visitor<V: EvalAssignmentVisitor>(
+    configuration: Option<&Configuration>,
+    visitor: &mut V,
+    subject_key: &Str,
+    subject_attributes: &Arc<Attributes>,
+    now: DateTime<Utc>,
+) -> HashMap<String, Result<Option<Assignment>, EvaluationError>> {
+    configuration.map_or_else(HashMap::new, |config| {
+        config
+            .flags
+            .compiled
+            .flags
+            .keys()
+            .map(|flag_key| {
+                let result = get_assignment_with_visitor(
```

Comment:
Pardon my lack of rust knowledge / memory of this codebase, but assuming graceful mode is on is there a top-level try catch here so an error getting one assignment won't tank all? (Or do we want it to tank all? Thinking not)

Reply:
All questions welcome thanks for bringing it up! This is not the public method available through the rust sdk so the graceful mode is not relevant - `eppo_core` is an internal package.

The `Result` enum in rust is the canonical way to return a tuple of `(desired, exceptional)` behavior which I am employing here:

The return is `HashMap<String, Result<Option<Assignment>, EvaluationError>>` and evaluation will happen against all flags - the return will be each flag will have a `Result`

My thinking then is the caller can decide whether what to do if any fail. In the edge function I am writing it will return an HTTP exception if any assignment fails - https://github.com/Eppo-exp/eppo-multiplatform/pull/69/files#diff-ff0287fa81e178dd11d342cf54ba1519848b1a29bf0b06192fc0aaec64d0120bR138

Reply:
A couple of notes and nits.

Rust doesn't have exception, so no try-catch, and no "non-graceful" mode. You can usually tell where the error boundary lies from the return type (where `Result` is). `HashMap<String, Result<Option<Assignment>, EvaluationError>>` for this functions means each hash map value would be its own result.

For this use case, it's actually better to merge this code with the calling site, so it could control that boundary. Right now, it continues to evaluate all assignments even if the first one failed. For the edge function, it's better to short-circuit and stop evaluating the rest of the flags if we're going to return error anyway.

nit: `Result` is not a tuple because tuples contain **both** values simultaneously. Enums contain only one of the two types (either-or). Mathematically, `struct`s behave as `AND` or `*` (multiplies the number of available values in the type), and enums in Rust are `OR` or `+` (number of available values is the sum of two domains). Which is why you may see this feature called "algebraic data types."

Reply:
Thinking about it now, I would like to challenge the short-circuiting behavior. We usually default to returning default value if evaluation failed for any reason. It seems strange that we decide to abort evaluation for edge function and return a global error for all flags, which could cause SDKs to error out on all evaluations even if the failed flag is never used.

Reply:
Yea your thinking makes sense @rasendubi - we could deliver as many flags as we have available which does seem better than none.

To support this should I keep the return type as `HashMap<String, Result<Option<Assignment>, EvaluationError>>` to be able to collect success and failures independently? I assume if I wrap the entire thing in a `Result` that is more of the binary "everything failed or everything was successful".